server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/rust/RustLanguagePlugin.kt

```diff
@@ -0,0 +1,175 @@
+        // TODO
+    }
+
+    override fun resolveModule(targetInfo: TargetInfo): RustModule? =
+        targetInfo.getRustCrateInfoOrNull()?.let { rustCrateInfo ->
+            val location = resolveTargetLocation(rustCrateInfo)
+            val crateRoot = resolveTargetCrateRoot(rustCrateInfo, location)
+            val dependencies = resolveTargetDependencies(rustCrateInfo)
+
+            RustModule(
+                crateId = rustCrateInfo.crateId,
+                location = location,
+                fromWorkspace = rustCrateInfo.fromWorkspace,
+                name = rustCrateInfo.name,
+                kind = rustCrateInfo.kind,
+                edition = rustCrateInfo.edition,
+                crateFeatures = rustCrateInfo.crateFeaturesList,
+                dependencies = dependencies,
+                crateRoot = crateRoot,
+                version = rustCrateInfo.version,
+                procMacroArtifacts = rustCrateInfo.procMacroArtifactsList,
+                procMacroSrv = rustCrateInfo.procMacroSrv,
+                rustcSysroot = rustCrateInfo.rustcSysroot,
+                rustcSrcSysroot = rustCrateInfo.rustcSrcSysroot,
+                cargoBinPath = rustCrateInfo.cargoBinPath,
+                rustcVersion = rustCrateInfo.rustcVersion,
+                rustcHost = rustCrateInfo.rustcHost,
+            )
+        }
+
+    private fun resolveTargetDependencies(rustCrateInfo: RustCrateInfo): List<RustDependency> =
+        rustCrateInfo.dependenciesList.map { depInfo ->
+            RustDependency(
+                crateId = depInfo.crateId,
+                rename = depInfo.rename,
+            )
+        }
+
+    private fun resolveTargetLocation(rustCrateInfo: RustCrateInfo): RustCrateLocation =
+        when (rustCrateInfo.location) {
+            BspTargetInfo.RustCrateLocation.WORKSPACE_DIR -> RustCrateLocation.WORKSPACE_DIR
+            else -> RustCrateLocation.EXEC_ROOT
+        }
+
+    private fun resolveTargetCrateRoot(rustCrateInfo: RustCrateInfo, location: RustCrateLocation): String {
+        val path = if (location == RustCrateLocation.WORKSPACE_DIR) {
+            bazelPathsResolver.relativePathToWorkspaceAbsolute(rustCrateInfo.crateRoot)
+        } else {
+            bazelPathsResolver.relativePathToExecRootAbsolute(rustCrateInfo.crateRoot)
+        }
+
+        return bazelPathsResolver.resolveUri(path).toString()
+    }
+
+    /**
+     * Find all targets that are needed to be resolved.
+     * It includes all targets from `targets` and all their dependencies.
+     *
+     * Let's assume we have a following project:
+     *           B    A
+     *           |   / \
+     *           C  D   E
+     *           \ /    |
+     *            F     G
+     * We want to resolve all targets related to B. We only need B, C and F.
+     *
+     * There is no need to revisit already visited targets - we can use `visited` set for that.
+     * */
+    private fun findAllRelatedRustTargets(
```

Comment:
cant u use `DependencyTree` here? (prob u need to create another instance)

Reply:
We could, but besides creating an instance of `DependencyTree` we would have to crate `TargetInfo`'s from all `Module`s and `Label`s. Will it be still better than what we have now?